bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -4481,6 +4476,9 @@ ReplicaImp::ReplicaImp(bool firstTime,
         KeyExchangeManager::instance().sendInitialKey(this, lastExecutedSeqNum);
         LOG_INFO(GL, "Send initial key exchange after completing state transfer " << KVLOG(lastExecutedSeqNum));
       }
+    } else {
+      LOG_WARN(GL,
```

Comment:
It can be written in a better way and maybe not start with the negated condition.
A replica should publish (one of) its public key(s) before other replicas can communicate with it, but this code suggests that it might have triggered state transfer before publishing its key. Once the replica returns to the quorum (meaning after it is done with state transfer) its key needs to be known to the other replicas.